antlr_ast/ast.py

```diff
@@ -283,3 +176,505 @@ def reportContextSensitivity(
+        if isinstance(result, AttributeError):
+            raise AttributeError('{}.{} is invalid.'.format(self.__class__.__name__, name))
+
+        return result
+
+    @classmethod
+    def combine(cls, *fields):
+        """Combine fields
+
+        Creates a list field from other fields
+        Filters None and combines other elements in a flat list
+        Use in transformer methods.
+        """
+        result = reduce(cls.extend_node_list, fields, [])
+
+        return result
+
+    @staticmethod
+    def extend_node_list(acc, new):
+        """Extend accumulator with Node(s) from new"""
+        if new is None:
+            new = []
+        elif not isinstance(new, list):
+            new = [new]
+        return acc + new
+
+    def isinstance(self, class_name):
+        return type(self).__name__ == class_name
+
+    def get_text(self, full_text=None):
+        # TODO implement as __str__?
+        #  + easy to combine with str/Terminal
+        #  + use Python instead of custom interface
+        # (-) very different from repr / json
+        if isinstance(self._ctx, ParseTree):
+            if full_text is None:
+                text = self._ctx.getText()
+            else:
+                text = full_text[self._ctx.start.start : self._ctx.stop.stop + 1]
+        else:
+            text = None
+
+        return text
+
+    def get_position(self):
+        ctx = self._ctx
+        d = {
+            "line_start": ctx.start.line,
+            "column_start": ctx.start.column,
+            "line_end": ctx.stop.line,
+            "column_end": ctx.stop.column + (ctx.stop.stop - ctx.stop.start),
+        }
+        return d
+
+    def to_json(self):
+        return {
+            "@type": self.__class__.__name__,
+            "@fields": self._fields,
+            "field_references": self._field_references,
+            "label_references": self._label_references,
+            "children": self.children,
+        }
+
+    def __repr__(self):
+        return str({**self.children_by_field, **self.children_by_label})
+
+
+# TODO:
+AstNode = BaseNode
+
+
+class Terminal(BaseNode):
+    """This is a thin node wrapper for a string.
+
+    The node is transparent when not in debug mode.
+    In debug mode, it keeps the link to the corresponding ANTLR node.
+    """
+
+    _fields = tuple(["value"])
+    DEBUG = False
+    DEBUG_INSTANCES = []
+
+    def __new__(cls, *args, **kwargs):
+        instance = super().__new__(cls, *args, **kwargs)
+        if cls.DEBUG:
+            cls.DEBUG_INSTANCES.append(instance)
+            return instance
+        else:
+            return args[0][0]
+
+    def __eq__(self, other):
+        return self.value == other
+
+    def __str__(self):
+        # currently just used for better formatting in debugger
+        return self.value
+
+    def to_json(self):
+        return str(self)
+
+    def __repr__(self):
+        return "'{}'".format(self.value)
+
+
+class AliasNode(BaseNode, metaclass=AstNodeMeta):
```

Comment:
AFAICT 

```
class AliasNode(BaseNode, AstNodeMeta):
```

would work fine (it looks like AstNodeMeta is more like a mixin; it doesn't do any work when a new class is defined)